### ISSUE TITLE
EasyPerform() - CURL error [60]=SSL peer certificate or SSH remote key was not OK

### ISSUE THREAD

**Issue by FranceAS89** [googleapis/google-cloud-cpp#6212]
**Does this issue affect the google-cloud-cpp project?**
No

**What component of google-cloud-cpp is this related to?**
something in google/cloud/storage

**Describe the bug**
I've built the whole project on Windows 8.1 using cMake 3.19.4 and Visual studio 16 2019

**set the environment variable:**
GOOGLE_APPLICATION_CREDENTIALS=[path]/gcp-ml-access.json

The minimal "quickstart" app describe here https://googleapis.dev/cpp/google-cloud-storage/latest/ fails with the following:

**ERROR:**
Error creating object: Permanent error in CreateResumableSession: EasyPerform() - CURL error [60]=SSL peer certificate or SSH remote key was not OK [UNKNOWN] [UNKNOWN]

I can use any name of "bucket" and the message is the same.

**Operating system**:
Windows 8.1

**What compiler and version are you using?**
The compiler: MSVC2019

**Comment by coryan** [comment]
Seems similar to #3770, the error indicates your libcurl+SSL libraries cannot verify the certificates coming from the server.  Typically this indicates missing root certificates or a curl library trying to use OpenSSL (and missing root certificates).

Most likely our documentation or build scripts need improvement, but I cannot figure out the problem without more information from you:

- What version of `libcurl` are you using?
- Was it compiled to use the native `schannel` or to use `OpenSSL`?
- If you have access to the `curl-config` binary, can you run `curl-config --ca`?

For my peace of mind (I know this does not help *you*), we do test the quickstart on Windows, and it works:

https://source.cloud.google.com/results/invocations/2f630a76-6399-46bb-a2c0-567323a7d8b8/targets/cloud-devrel%2Fclient-libraries%2Fcpp%2Fgoogle-cloud-cpp%2Fwindows%2Fquickstart-cmake-static/log

On the remote chance this helps, the driver scripts are:

https://github.com/googleapis/google-cloud-cpp/blob/master/ci/kokoro/windows/build-quickstart-cmake.ps1
https://github.com/googleapis/google-cloud-cpp/blob/master/ci/kokoro/windows/build-cmake-dependencies.ps1
https://github.com/googleapis/google-cloud-cpp/blob/master/ci/kokoro/windows/build.ps1

As I said above, this is not to say you are doing something wrong, just that our instructions or build scripts are not precise / detailed / accurate enough.

**Comment by FranceAS89** [comment]
I answer your questions:
- libcurl version is 7.76.0
- to use OpenSSL 1.0.2
- i run curl-config --ca, but I obtain this screen
![image](https://user-images.githubusercontent.com/82205059/114201295-53f90d00-9956-11eb-9ab2-307030580b62.png)
Thanks for your help.

**Comment by coryan** [comment]
> I answer your questions:
> 
> * libcurl version is 7.76.0
> * to use OpenSSL 1.0.2

Okay... Please see this for background:

https://curl.se/docs/sslcerts.html

At the moment, on Windows, if you are not using Schannel, `google-cloud-cpp` requires that you configure the location of the CA info. You can use this flag:

https://github.com/googleapis/google-cloud-cpp/blob/cb87d6a3d2847d63e01283228e4749c786f3a46f/google/cloud/storage/client_options.h#L123
https://github.com/googleapis/google-cloud-cpp/blob/cb87d6a3d2847d63e01283228e4749c786f3a46f/google/cloud/storage/client_options.h#L290

https://curl.se/libcurl/c/CURLOPT_CAINFO.html

In passing: please consider updating your OpenSSL version, 1.0.2 is fairly old, it has problems in multi-threaded environments, something in the OpenSSL 1.1 series would be a good start.

In addition, TIL about `CURLSSLOPT_NATIVE_CA`: 

https://stackoverflow.com/questions/37551409/configure-curl-to-use-default-system-cert-store-on-windows
https://curl.se/libcurl/c/CURLOPT_SSL_OPTIONS.html

Seems like a useful feature to turn on if one has a recent version of libcurl.  I created #6214, that would have prevented this problem.

> * i run curl-config --ca, but I obtain this screen

Yeah, seems like this is not installed / compiled for Windows. My bad.

**Comment by FranceAS89** [comment]
I tryed to follow your advice.
First, I changed quickstart and I added the following lines:
![image (2)](https://user-images.githubusercontent.com/82205059/114214610-cc66ca80-9964-11eb-81bb-9540e9ca61e9.png)
I downloaded cacert.pem from https://curl.se/docs/caextract.html
Then I changed the following file ..\cloud-library\curl-master\lib\doh.c:
![image (3)](https://user-images.githubusercontent.com/82205059/114214954-49923f80-9965-11eb-8000-fcc709a11371.png)
Despite both changes I received the same before error.

**Comment by FranceAS89** [comment]
I was wondering if the problem was due to our Google project settings. Could you check if my credentials works on your pc? If you are agree, I could send you in private my file.json and bucket name.

**Comment by coryan** [comment]
> I was wondering if the problem was due to our Google project settings. Could you check if my credentials works on your pc? If you are agree, I could send you in private my file.json and bucket name.

The error you are getting is before Google gets to validate your credentials.  It is your client validating Google's certificate.

If you want to perform a test like that, I would recommend you use something like `gcloud` or `gsutil` locally, without having to share your credentials.

**Comment by coryan** [comment]
> First, I changed quickstart and I added the following lines:
> ![image (2)](https://user-images.githubusercontent.com/82205059/114214610-cc66ca80-9964-11eb-81bb-9540e9ca61e9.png)

Did you also pass that parameter to initialize a `gcs::ClientOptions` object and pass that to `gcs::Client`?  What I mean is you need to change this:

https://github.com/googleapis/google-cloud-cpp/blob/cb87d6a3d2847d63e01283228e4749c786f3a46f/google/cloud/storage/quickstart/quickstart.cc#L32

to say something like:

```cc
  auto const path = std::string{"..." /* your long path here */};

 StatusOr<gcs::Client> client = gcs::Client(gcs::ClientOptions::CreateDefaultClientOptions(
      gcs::ChannelOptions{}.set_ssl_root_path(path)).value());
```

Note that my use of `.value()` there can result in an exception, I am trying to simplify the code for this question, you may want to check its value before proceeding.

In any case.  Seems like we would need to reproduce this locally.  Would you be able to tell me more about how exactly you compiled the `google-cloud-cpp` code and its dependencies? All the commands / steps would be ideal, a general description if you cannot provide details would be a good start.

> I downloaded cacert.pem from https://curl.se/docs/caextract.html
> Then I changed the following file ..\cloud-library\curl-master\lib\doh.c:
> ![image (3)](https://user-images.githubusercontent.com/82205059/114214954-49923f80-9965-11eb-8000-fcc709a11371.png)
> Despite both changes I received the same before error.

Out of curiousity, have you tried to use the `curl` program in `curl-master` to download a file?  Say https://pki.google.com/roots.pem?  I think the command would be something like:

```cmd
cd c:\Develop\Packages\cloud-library\curl-master\install_win64\bin
curl.exe -sSL https://pki.google.com/roots.pem
```

Until you get that to work, it is unlikely that the `google-cloud-cpp` library will work.

**Comment by FranceAS89** [comment]
Great. I changed code as you said, now problem is:
![image (5)](https://user-images.githubusercontent.com/82205059/114221659-de00a000-996d-11eb-8560-6cd4be9528aa.png)
I think that problem is due to my Google Service account permission, but my account service is already the owner's project. What dou you think about it? Next week, I'll continue to work about it.

**Comment by coryan** [comment]
Glad that worked.  You should verify the `logica3` is a bucket and that you have permissions for it, something like:

```
gsutil ls -b -L gs://logica3
```

Please keep in mind that buckets are in a global namespace... if any other project has created a bucket called `logica3` you will not be able to create objects on it (or access the bucket):

https://cloud.google.com/storage/docs/naming-buckets

**Comment by MrMoose** [comment]
Hey, that's interesting. I just ran into the exact same problem in my Windows build. @coryan , thanks for the heads up! Until now I bypassed the problem by switching off peer verification but after reading this I have rebuilt libcurl to use Schannel and removed the OpenSSL and libssh2 dependencies.

After that, relink into lib google-cloud-sdk and now it works. Well, at least Storage does but I'll be able to confirm other ones like pubsub soonish.

So, bottom line is, build your libcurl with Schannel and _no_ OpenSSL or SSH2.

A bit strange though. I've been using lots of libs that use OpenSSL and CURL for such purposes and never encountered that problem. Is there anything strange about the root certificates being used by Google?

Cheers,
Moose
